grudge/op.py

```diff
@@ -310,6 +326,82 @@ def mass(dcoll, *args):
     return _bound_mass(dcoll, dd)(u=vec)
 
 
+def reference_mass_matrix(actx, out_element_group, in_element_group):
+    @keyed_memoize_in(
+        actx, reference_mass_matrix,
+        lambda out_grp, in_grp: (out_grp.discretization_key(),
+                                 in_grp.discretization_key()))
+    def get_ref_mass_mat(out_grp, in_grp):
+        if out_grp == in_grp:
+            from meshmode.discretization.poly_element import mass_matrix
+
+            return actx.freeze(
+                actx.from_numpy(mass_matrix(out_grp))
+            )
+
+        from modepy import vandermonde
+        basis = out_grp.basis_obj()
+        vand = vandermonde(basis.functions, out_grp.unit_nodes)
+        o_vand = vandermonde(basis.functions, in_grp.unit_nodes)
+        vand_inv_t = np.linalg.inv(vand).T
+
+        weights = in_grp.quadrature_rule().weights
+        return actx.freeze(
+            actx.from_numpy(
+                np.einsum("j,ik,jk->ij", weights, vand_inv_t, o_vand)
+            )
+        )
+
+    return get_ref_mass_mat(out_element_group, in_element_group)
+
+
+def _apply_mass_operator(dcoll, dd_out, dd_in, vec):
+    from grudge.geometry import area_element
+
+    in_discr = dcoll.discr_from_dd(dd_in)
+    out_discr = dcoll.discr_from_dd(dd_out)
+
+    actx = vec.array_context
+    area_elements = area_element(actx, dcoll, dd=dd_in)
+    return DOFArray(
+        actx,
+        tuple(actx.einsum("ij,ej,ej->ei",
+                          reference_mass_matrix(
+                              actx,
+                              out_element_group=out_grp,
+                              in_element_group=in_grp
+                          ),
+                          ae_i,
+                          vec_i,
+                          arg_names=("mass_mat", "jac_det", "vec"),
+                          tagged=(MassOperatorTag(),),
+                          tagged_array_axes={
+                              "mass_mat": "stride:auto,stride:auto"
+                           })
```

Comment:
@inducer I managed to get the mass matrix test to finally work. Requires: https://github.com/inducer/meshmode/pull/170

I'm not 100% sure this is the best way to do this, as this was a rather quick attempt to get something working. To be perfectly honest, I'm still not 100% sure if there are any unknown consequences for choosing these parameters.

I read up the docs [here](https://documen.tician.de/loopy/ref_transform.html?highlight=tag_array_axes#loopy.tag_array_axes) but they weren't extremely illuminating. I dug around a bit in the code [here](https://github.com/inducer/loopy/blob/main/loopy/kernel/array.py#L210-L417) (the docs suggested [here](https://documen.tician.de/loopy/ref_internals.html#loopy.kernel.array.parse_array_dim_tags) but there's nothing there, so I went straight to the source code) and it was even less obvious. I didn't see a summary describing the acceptable parameters, so I dug line-by-line to back out what acceptable parameters were (as well as taking some hints from some examples in the test suite).

Maybe I wasn't looking in the right place.

Reply:
Remind me sometime to show you how to look at the generated CL code, to make all this a bit less mysterious. (In fact, I'll just add that to our agenda for tomorrow.)